### PR TITLE
Integrate Peer Network Interactions into Skunk Application

### LINKED ISSUE #31: Connect Peer Network Code to Skunk Main Application Logic
### Integrate Asynchronous Peer Network Interactions in Skunk Application

**Description:**  
We need to establish a connection between the peer network code and the Skunk application logic using the `networkInteractionPort` and the `NetworkConnection` interface. Relevant source code is located at [networkInteractionPort.go](https://github.com/scherzma/Skunk/blob/86200c9205716b990e838127b783293f5e693c33/cmd/skunk/application/port/network/networkInteractionPort.go).

**Requirements:**
- Implement asynchronous interactions between the peer network and Skunk application logic to allow users to read and send messages simultaneously.
- The user's message input via the frontend should be routed through the main application logic and then sent to the peer network using `SendMessageToNetworkPeer`.
- The application logic needs to initially subscribe to the peer network with `SubscribeToNetwork` and should be able to unsubscribe using `UnsubscribeFromNetwork`.
- Consider adapting some methods to support asynchronous message exchanges.

**Tasks:**
- [x] Review the current implementation of `NetworkConnection` in the `networkInteractionPort`.
- [x] Modify methods as necessary to support asynchronous operations.
- [x] Implement message handling workflow that includes subscribing, sending messages, and unsubscribing.

### PR DESCRIPTION
**Pull request by JavaHammes**
This pull request implements the necessary networkInteraction port and thus connects the app logic and the network logic. A networkAdapter was developed for this purpose, which offers the following functionalities:
- `SubscribeToNetwork(observer network.Observer)`
    - Starts and connects the app logic to the network 
- `UnsubscribeFromNetwork()`
    - Terminates and disconnects the Logic app from the network
- `SendMessageToNetworkPeer(address string, message network.Message)`
    - Sends a message to another network peer

In addition, errors in the network peer were fixed and the heartbeat mechanism was removed as it was too error-prone. 

Closes #31

### CODE REVIEW

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+// SendMessageToNetworkPeer sends a message to a specified network peer
+func (n *NetworkAdapter) SendMessageToNetworkPeer(address string, message network.Message) error {
+	// connect to peer if not already connected
+	if !n.peer.IsConnectedTo(address) {
+		err := n.peer.Connect(address)
+		// if there is any error, we treat it as if the peer is offline
+		if err != nil {
+			// message subscriber that the peer is offline
+			message := network.Message{
```
Comment (scherzma) [inline cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:121]: The network.Message struct will get changes in one of the next pull requests. This works for now, but will have to be adjusted in the future.

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	UseEmbedded          = true // use embedded tor process
+)
+
+var (
+	networkAdapter *NetworkAdapter
+	once           sync.Once
+)
+
+// NetworkAdapter connects the main logic to the tor peer network
+type NetworkAdapter struct {
```
Comment (scherzma) [inline cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:33]: Should that be "networkAdapter"? To be private.

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+	n.subscriber = observer
+	n.peer = peerInstance
+	n.tor = torInstance
+
+	// notify the subscriber that the network is now online and send the onion address
+	message := network.Message{
+		Id:              util.UUID(),
+		Timestamp:       util.CurrentTimeMillis(),
+		Content:         n.peer.Address,
```
Comment (scherzma) [inline cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:78]: We can send it as a normal string, but so far everything in a messageContent is JSON. Maybe we should also make this JSON.

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+	if n.peer == nil {
+		return fmt.Errorf("peer network is nil")
+	}
+
+	if n.tor == nil {
+		return fmt.Errorf("tor network is nil")
+	}
+
+	// stop tor and peer services
+	stopTor(n.tor)
```
Comment (scherzma) [inline cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:104]: Why not just n.tor.StopTor()?

```
// stopTor stops the tor service
func stopTor(torInstance *tor.Tor) {
	torInstance.StopTor()
}
```

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+	stopPeer(n.peer)
```
Comment (scherzma) [inline cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:105]: Same as above.

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+// startPeer initializes a peer network connection using the provided OnionService
+func startPeer(onionService *cretztor.OnionService) (*peer.Peer, error) {
+	peerInstance, err := peer.NewPeer(onionService.ID+".onion", LocalPort, RemotePort, "127.0.0.1:"+SocksPort)
+	if err != nil {
+		return nil, err
+	}
+	peerInstance.Listen(onionService)
+
+	time.Sleep(1 * time.Second)
```
Comment (scherzma) [inline cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:237]: Is time.Sleep the best solution?

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+	return peerInstance, err
+}
+
+// stopTor stops the tor service
+func stopTor(torInstance *tor.Tor) {
```
Comment (scherzma) [inline cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:243]: Is this needed?

File: cmd/skunk/adapter/in/peer/peer.go [hunk 8]
```diff
@@ -219,23 +215,33 @@ func (p *Peer) readMessage(conn *websocket.Conn, address string) (string, error)
 
 // ReadMessages starts readMessage for every conn in readConns in the readRate interval.
 func (p *Peer) ReadMessages(messageCh chan<- string, errorCh chan<- error) {
```
Comment (scherzma) [inline cmd/skunk/adapter/in/peer/peer.go:217]: Maybe add a few comments.

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go (hunk 6 above)
Comment (JavaHammes) [reply cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:237]: no, i just didn't have a better solution. but now we have.

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go (hunk 4 above)
Comment (JavaHammes) [reply cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:104]: i thought that there might be more functionallity to add later (see `StopTor` in tor.go). But I agree, right now it is not really necessary.

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go (hunk 5 above)
Comment (JavaHammes) [reply cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:105]: See other comment

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go (hunk 7 above)
Comment (JavaHammes) [reply cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:243]: See other comment

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go (hunk 2 above)
Comment (JavaHammes) [reply cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:33]: Why should that be private?

File: cmd/skunk/adapter/in/networkAdapter/networkAdapter.go (hunk 3 above)
Comment (JavaHammes) [reply cmd/skunk/adapter/in/networkAdapter/networkAdapter.go:78]: Done in: [ea02765](https://github.com/scherzma/Skunk/pull/33/commits/ea027653036eff1b5fd8c22e4d9656ae0a8c2187)

File: cmd/skunk/adapter/in/peer/peer.go (hunk 8 above)
Comment (JavaHammes) [reply cmd/skunk/adapter/in/peer/peer.go:214]: done in: [130f30a](https://github.com/scherzma/Skunk/pull/33/commits/130f30aca6934913c3b65c19c3710094c3592fd3)

**Review by scherzma — APPROVED** [review]
LGTM! nice nice

**Pull request merged**